### ISSUE TITLE
Create a gomega-community repo?

### ISSUE THREAD

**Issue by chendrix** [onsi/gomega#165]
I'm thinking about writing a gomega matcher for https://github.com/pkg/errors, and I'm looking at the canonical place to contribute this back to the community.

There's protobuf matchers and yaml matchers that require third party libraries inside the `matchers` package, and there's also things like https://github.com/tjarratt/gcounterfeiter which move custom matchers to their own repo.

I like the idea of it being separate from base gomega, but I think having repos for 1 file is a bit overkill.

What are people's thoughts on creating a `gomega-community` repo with separate subpackages for each custom matcher, where people can contribute via pull requests? Then each matcher can also have its own licensing and maintain atomic importability?

**Comment by onsi** [comment]
I like it - options include:
- a separate repo (happy to start github.com/onsi/gomega-community)
- a community folder within Gomega itself?  or would that make the licensing concerns complex?

**Comment by chendrix** [comment]
I would say a separate repo, with no top-level License.

**Comment by tjarratt** [comment]
I'm all for keeping third-party-aware matchers out of Gomega core, but I don't know if I understand the exact pain we're trying to solve.

Is the goal to move those matchers out of Gomega to resolve the pain in installing gomega? (I'm for this, and I think @robdimsdale   and @onsi are too).

Or is the goal to to aid in discoverability of matchers? I purposefully left mine in my own repo because I didn't want to pollute `onsi/gomega`. Having a global catch-all place for Gomega matchers feels like we'll have the same pain we feel today anytime you want to bring in a few extra matchers. I worry this will just become a dumping ground for matchers and it will discourage users from only bringing in what they need for testing.

Could we solve the problem by moving the `protobuf` and `yaml` matchers out of `gomega` and just have a list of links to 3rd party matchers (in the README and docs maybe?).

**Comment by tjarratt** [comment]
Or am I wrong and people want to have a global repo with various matchers in it? (I could very well be in the minority here, since I've been out of the go-lang game for a few months now).

**Comment by onsi** [comment]
So many the options!

Option 1:
- new gomega-community repo, folks put things there if they'd like to
- obviously folks can maintain their own repos too
- a list of community matchers in github.com/onsi/gomega with links to the relevant repos

Option 2:
- new gomega-community github *org*, put things there
- obviously folks can maintain their own repos too but would be invited to contribute to the org
- a list of community matchers in github.com/onsi/gomega with links to the relevant repos

Option 3:
- a folder in gomega
- a list of community matchers in github.com/onsi/gomega with links to the relevant repos

**Comment by onsi** [comment]
Also - how many folks wanna see the yaml and protobuf matchers moved out of gomega core?

👍 for move em out
👎 for keep em in

**Comment by tjarratt** [comment]
My vote is for the following

* Move `yaml` and `protobuf` out of gomega core.
* Option 2 : new gomega-community github _org_

**Comment by Amit-PivotalLabs** [comment]
Could anyone speak to why they want to move `yaml` and `protobuf` matchers out?  Are the reasons the same for both matchers?

Our team maintains its own YAML matcher because there was none in gomega at the time.  After writing it, we offered to contribute it in, but were told for some reason I can't recall (licensing, introducing dependencies?) that gomega didn't want to do that.  We saw later that a different YAML matcher was added anyways.  We took that as a sign that at some point when we get a chance, we could/should go ahead and contribute our matcher back to gomega since it was an improvement on the current YAML matcher in gomega.

A couple reasons for wanting to do this were:

1. have there be just 1 canonical YAML matcher, so that new folks on our team aren't confused about why we have a YAML matcher when gomega already has one
2. offload maintenance to gomega, since the core focus of our team is not this gomega matcher, but it's still something we have to maintain CI for in theory

So for our needs, having something either in gomega core or at least a folder in the gomega repo (I'm not really sure what the difference is, except you could in theory import less if you just import base matchers and not the community matchers?) would be the best.

Hence I'm curious why the desire to remove those other matchers out of core, and whether moving them out of core but still in the gomega repo solves people's pains.

/cc @azankich

**Comment by xoebus** [comment]
I didn't even know about the YAML one but the dependency on the Protocol Buffer libraries when most projects don't use them is a shame.

**Comment by Amit-PivotalLabs** [comment]
That's what I would have guessed.  Which makes me think that the reasons might be different for the two matchers.

**Comment by chendrix** [comment]
My proposed organization of the `gomega-community` repo would be separate packages inside of it for each of the matchers, e.g. github.com/chendrix/gerrors could be at `github.com/onsi/gomega-community/pkgerrors`, the yaml matcher at `github.com/onsi/gomega-community/yaml`. That way there's a segregation of import paths and you can atomically import only the matchers that you wanted. There can also be separate LICENSE/README for each subpackage.

Having one repo does aid in discoverability. I would have never known about github.com/tjarratt/gcounterfeiter if I wasn't closer to the community, but it's actually an *extremely* valuable library.

I don't mind an org instead, just don't personally know the logistics around creating/maintaining them. I think one downside to them is a proliferation of extremely tiny repositories, with potentially a lot of independent/individual noise splitting issues across every repo? May also still want a one-pager listing all of the various matchers available instead of needing to search the github org.

**Comment by tinygrasshopper** [comment]
**tl;dr** I like that all the matchers are in one place, cause they are easy to discover and consume 

I am trying to understand which problems are we trying to solve with the extraction of `gomega-community` repo. I see two problems being discussed. 

1. Adding more dependencies to gomega, itself. 
1. Difficulty in installing gomega with `go get`

For number 1, I think its fair that gomega pulls some dependencies, as it would be included as a test dependecy in its consumers. One downside can be the dependencies can have licences which are not compatible with gomega's licence. Which has happened in the past, we have to be mindful while adding newer dependencies. 

For number 2, we did discuss this pain point, and I agree it sucks. But we have identified a solution for it in https://github.com/onsi/gomega/issues/139, but I haven't gotten to making the change yet. 

I think splitting the matchers in a seperate repo would increase the cost of discovery of the matchers. The way you consume 'community matchers' would impact usablilty as well. The current usability flow for the `yaml` matcher for example is, `go get github.com/onsi/gomega`, add `import . github.com/onsi/gomega` and use the matcher. For the `gomega-community` matcher it would be`go get github.com/onsi/gomega`, `go get github.com/onsi/gomega-community/yaml` and adding dot imports for `github.com/onsi/gomega` and `github.com/onsi/gomega-community/yaml`. 


btw, @chendrix, the errors matcher you have written in https://github.com/chendrix/gerrors, doesn't have a code depedency on `github.com/pkg/errors`, it has a test dependecy. If we add it to gomega `go get` won't fetch the `github.com/pkg/errors`. Also, I think it would be a useful addition to gomega.

**Comment by tedsuo** [comment]
The problem with a `gomega-community` organization is that it creates a set of blessed community resources. That impedes organic evolution, and requires a curator to pay attention to it. There's a long history of this kind of community solution falling into disrepair over time.

The standard community solution would be a `Ginkgo/Gomega Registry` where users can register plugins which they maintain in their own repos. This leaves things open, so chendrix can have an `errors` matcher and I can have a `gerrors` matcher and the community can evolve from using one to the other when chendrix wanders away from golang, or use both if they have different approaches that appeal to different people. 

The obvious downside with a registry is it would require software to be built and maintained. If no one has the appetite for this, I'm willing to concede that a `gomega-community` repo is possibly a "worse is better" solution that solves this problem in the short run. But it will fail at scale, as it has traditionally done so in other communities.

**Comment by onsi** [comment]
A few things:

1. I'm seeing more folks wanting to pull the yaml and protobuf matchers out.  This will be a breaking change but may be worth it in the long run.
2. We want a way for folks to discover "community" matchers.

For 1 these can just get pulled out and put under github.com/onsi/gomega-proto and /gomega-yaml?

For 2 the *simplest* possible solution would be a list of community matchers/libraries in the README.  Solves discoverability and contributing is a simple PR to gomega.  No?  At PR time we could have a discussion around whether or not we'd like to pull the matcher into core vs have it remain separate.  Thoughts?

We can start with 2 right now and continue to hash out 1 on this thread.

**Comment by tjarratt** [comment]
That's fascinating, Ted. I haven't seen that play out in other languages or matcher frameworks (yet? 😁). 

Can you give an example of where that's occurred? That's a really compelling story you told that's helping guide me (at least) towards something besides a separate org or repo. 

Sent From A Very Small Keyboard

**Comment by luan** [comment]
Sorry I'm a little late to the party here.

I agree with @tedsuo's point about the `gomega-community` org or repo. That is, in my opinion, specifically **not** community, since it's maintained by a specific team it closes the door for community members to really contribute their own matchers. I feel like the barrier of entry is too high:

- I find a missing matcher
- I write the matcher
- I send a PR to `gomega-community`
- We bikeshed over the matcher (because we will maybe not want a specific matcher or a style of matching that it uses, or failure message)
- It gets merged in (or worse, it doesn't)

I would feel strongly discouraged go through that to contribute with a matcher.

I like the idea of a registry of matchers/addons for gomega, but I feel like a live registry (saas style) is overkill (at least right now), the README list is a valid implementation of a registry, especially if we make that really discoverable and easy to contribute to (adding a matcher/addon to the list should be trivial, via PR):

- I find a missing matcher
- I write the matcher
- I push it to my repo

Now at that step I can decide if I'm ready for others to use it, I'll start using it right away. There's a race here, much like with @Amit-PivotalLabs' example with the YAML matcher, someone might implement the same idea. That's OK, we can have addons that fulfill the same purpose on the list.

When I'm ready I request that my matcher is added to the list. It gets merged quickly (we should make sure it at least has enough test coverage for us to be recommending it, but it should not be a huge deal).

As far as solving the immediate problem with proto and YAML matchers. I'd say copy them out (probably to gomega-proto/gomega-yaml -- probably under @onsi's account?), add a deprecation to the current ones (that hopefully tells you somehow that you should update things), and delete them in a couple of months.

**Comment by BooleanCat** [comment]
As requested - when you add the list of community matchers/utils, please include [gfile](https://github.com/BooleanCat/gfile).

**Comment by chendrix** [comment]
To update everyone, I've generated a page on the [gomega wiki](https://github.com/onsi/gomega/wiki) with a list of the currently known Community matchers

Feel free to edit this as you will.

**Comment by onsi** [comment]
Sweet - gonna close this out.
